src/core/dbus-cgroup.c

```diff
@@ -303,6 +303,39 @@ static int property_get_ip_address_access(
         return sd_bus_message_close_container(reply);
 }
 
+static int property_get_disable_controller(
+                sd_bus *bus,
+                const char *path,
+                const char *interface,
+                const char *property,
+                sd_bus_message *reply,
+                void *userdata,
+                sd_bus_error *error) {
+
+        CGroupContext *c = userdata;
+        CGroupController ctrl;
+        int r;
+
+        assert(bus);
+        assert(reply);
+        assert(c);
+
+        r = sd_bus_message_open_container(reply, 'a', "s");
+        if (r < 0)
+                return r;
+
+        for (ctrl = 0; ctrl < _CGROUP_CONTROLLER_MAX; ctrl++) {
+                if ((c->disable_controllers & CGROUP_CONTROLLER_TO_MASK(ctrl)) == 0)
+                        continue;
+
+                r = sd_bus_message_append(reply, "s", cgroup_controller_to_string(ctrl));
+                if (r < 0)
+                        return r;
+        }
+
+        return sd_bus_message_close_container(reply);
+}
```

Comment:
hmm, this is very very similar to property_get_delegate_controllers(). Which I think means this implementation should be generic, and hence work for any CGroupMask field. i.e. in the SD_BUS_PROPERTY() line use the offset field to point this call directly to `c->disable_controllers` rather than to `c` itself. This function should then be named something like `property_get_cgroup_mask()` or, and directly hooked up inside the SD_BUS_PROPERTY() line.

And then property_get_delegate_controllers() should be changed to first do the extra `!c->delegate` check and then simply call your new function for the rest of the work... Of course this would mean that the SD_BUS_PROPERTY() line for `DelegateControllers` should use a zero offset, and the offset would be added in in `property_get_delegate_controllers()` while the SD_BUS_PROPERTY() line for `DisableControllers` would directly use the right offset, if you follow what I am saying.